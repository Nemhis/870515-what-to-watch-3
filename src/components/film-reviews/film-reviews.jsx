import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import dayjs from 'dayjs';

import {getCommentsForSelectedFilm} from '../../reducer/operation/selectors';
import {isOdd} from '../../utils';


class FilmReviews extends PureComponent {
  componentDidMount() {
    this.props.onMount();
  }

  _renderReview({id, comment, rating, date, user}) {
    const dayjsDate = dayjs(date);

    return (
      <div className="review" key={id}>
        <blockquote className="review__quote">
          <p className="review__text">{comment}</p>

          <footer className="review__details">
            <cite className="review__author">{user.name}</cite>
            <time className="review__date" dateTime={dayjsDate.format(`YYYY-MM-DD`)}>
              {dayjsDate.format(`MMMM D, YYYY`)}
            </time>
          </footer>
        </blockquote>

        <div className="review__rating">{rating}</div>
      </div>
    );
  }

  _divideComments(comments) {
    const even = [];
    const odd = [];

    comments.forEach((review, index) => {
      if (isOdd(index)) {
        odd.push(review);
      } else {
        even.push(review);
      }
    });

    return [even, odd];
  }

  render() {
    const {comments} = this.props;
    const [leftColumn, rightColumn] = this._divideComments(comments);

    return (
      <div className="movie-card__reviews movie-card__row">
        {leftColumn.length &&
          <div className="movie-card__reviews-col">
            {leftColumn.map(this._renderReview)}
          </div>
        }

        {rightColumn.length &&
          <div className="movie-card__reviews-col">
            {rightColumn.map(this._renderReview)}
          </div>
        }
      </div>
    );
  }
}

FilmReviews.propTypes = {
  onMount: PropTypes.func.isRequired,
  comments: PropTypes.arrayOf(PropTypes.shape({
    rating: PropTypes.number.isRequired,
    comment: PropTypes.string.isRequired,
    date: PropTypes.object.isRequired,
    user: PropTypes.shape({
      name: PropTypes.string.isRequired,
    }).isRequired,
  })).isRequired,
};

const mapStateToProps = (state) => ({
  comments: getCommentsForSelectedFilm(state),
});

export {FilmReviews};
export default connect(mapStateToProps)(FilmReviews);
