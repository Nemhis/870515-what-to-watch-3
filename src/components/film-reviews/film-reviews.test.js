import React from 'react';
import renderer from 'react-test-renderer';

import {FilmReviews} from './film-reviews.jsx';
import Comment from '../../adapter/comment';

const comments = [
  new Comment({
    'id': 1,
    'rating': 8.1,
    'comment': `test comment`,
    'date': `2018-04-04T16:00:00.000Z`,
    'user': {
      'id': 2,
      'name': `User name`,
    },
  }),
];

it(`Render correctly <FilmReviews>`, () => {
  const tree = renderer
    .create(<FilmReviews onMount={() => {}} comments={comments}/>)
    .toJSON(`test`);

  expect(tree).toMatchSnapshot();
});
