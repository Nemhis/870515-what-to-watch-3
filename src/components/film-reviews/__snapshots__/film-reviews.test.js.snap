// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly <FilmReviews> 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          test comment
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            User name
          </cite>
          <time
            className="review__date"
            dateTime="2018-04-04"
          >
            April 4, 2018
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        8.1
      </div>
    </div>
  </div>
  0
</div>
`;
